segmentation/include/pcl/segmentation/extract_clusters.h

```diff
@@ -41,10 +41,148 @@
+  * \param cloud the point cloud message
+  * \param additional_filter_criteria functor which is used as an additonal criteria that all points being added to the cluster must satisfy
+  * \param tree the spatial locator (e.g., kd-tree) used for nearest neighbors searching
+  * \note the tree has to be created as a spatial locator on \a cloud and \a indices
+  * \param tolerance the spatial cluster tolerance as a measure in L2 Euclidean space
+  * \param clusters the resultant clusters containing point indices (as a vector of PointIndices)
+  * \param min_pts_per_cluster minimum number of points that a cluster may contain (default: 1)
+  * \param max_pts_per_cluster maximum number of points that a cluster may contain (default: max int)
+  * \warning It is assumed that that the tree built on the same indices and cloud as passed here,
+  * if that is not the case then things can go very wrong. For speed reasons, we only check the sizes and not the content
+  * \ingroup segmentation
+  */
+  template <typename PointT, typename FunctorT> void
+   extractEuclideanClusters(
+      ConstCloudIterator<PointT> &it, const PointCloud<PointT> &cloud, FunctorT additional_filter_criteria,
+      const typename search::Search<PointT>::Ptr &tree, float tolerance, std::vector<PointIndices> &clusters,
+      unsigned int min_pts_per_cluster = 1, unsigned int max_pts_per_cluster = std::numeric_limits<int>::max())
+  {
+    if (tree->getInputCloud ()->points.size () != cloud.points.size ())
+    {
+      PCL_ERROR ("[pcl::extractEuclideanClusters] Tree built with a different point cloud size (%lu) than the input cloud (%lu)!\n", tree->getInputCloud ()->points.size (), cloud.points.size ());
+      return;
+    }
+
+    // Check if the tree is sorted -- if it is we don't need to check the first element
+    index_t nn_start_idx = tree->getSortedResults () ? 1 : 0;
+    // Create a bool vector of processed point indices, and initialize it to false
+    std::vector<bool> processed (cloud.points.size (), false);
+
+    for (; it.isValid(); ++it) {
+      if (processed[it.getCurrentIndex()])
+        continue;
+
+      index_t sq_idx = 0;
+      clusters.emplace_back();
+      auto& seed_queue = clusters.back();
+      seed_queue.indices.push_back (it.getCurrentIndex());
+
+      processed[it.getCurrentIndex()] = true;
+
+      while (sq_idx < static_cast<index_t> (seed_queue.indices.size()))
+      {
+        Indices nn_indices;
+        std::vector<float> nn_distances;
+
+        // Search for sq_idx
+        if (!tree->radiusSearch (seed_queue.indices[sq_idx], tolerance, nn_indices, nn_distances))
+        {
+          sq_idx++;
+          continue;
+        }
+
+        for (index_t j = nn_start_idx; j < nn_indices.size (); ++j) // can't assume sorted (default isn't!)
+        {
+          if (processed[nn_indices[j]]) // Has this point been processed before ?
+            continue;
+
+          if (additional_filter_criteria(it.getCurrentIndex(), j, nn_indices)) {
```

Comment:
I'd advice to put `j` after `nn_indices` to be consistent and to put `cloud` as the first parameter (to show `i` belongs to the cloud)